[3.12.x] Transition to schemas.metadata.geo.ca

jodygarnett opened the pull request:
This is a back-port of https://github.com/metadata101/iso19139.ca.HNAP/pull/323 to 3.12.x branch.

This pull request prepares the 3.12.x branch for the relocation of hnap schemas and registry to schemas.metadata.geo.ca.

For context and discussion see https://github.com/metadata101/iso19139.ca.HNAP/issues/118

jodygarnett committed:
Change codeList location to https://schemas.metadata.geo.ca/register/registerItemClasses-eng.html

jodygarnett committed:
Change schemaLocation to https://schemas.metadata.geo.ca/2009

jodygarnett committed:
Upgrade instructions for migration to schemas.metadata.geo.ca

ianwallen commented on src/main/plugin/iso19139.ca.HNAP/templates/hnaplevel1-english.xml:
```diff
@@ -119,9 +119,8 @@
-            <gmd:CI_RoleCode codeList="http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml#IC_90"
+            <gmd:CI_RoleCode codeList="https://schemas.metadata.geo.ca/register/napMetadataRegister.xml#IC_90"
                              codeListValue="RI_414">pointOfContact; contact</gmd:CI_RoleCode>
-         </gmd:role>
```
Is there a reason that you removed the ending element?

jodygarnett replied:
I do not believe this was intentional; it may of been a mistake in backport conflict resolution.

jodygarnett replied:
Actually Ian it is one line down, I expect this is a whitespace change.

ianwallen commented on src/main/plugin/iso19139.ca.HNAP/templates/hnaplevel1-english.xml:
```diff
@@ -361,9 +360,8 @@
-                              <gmd:CI_DateTypeCode codeList="http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml#IC_87"
+                              <gmd:CI_DateTypeCode codeList="https://schemas.metadata.geo.ca/register/napMetadataRegister.xml#IC_87"
                                                    codeListValue="RI_367">publication; publication</gmd:CI_DateTypeCode>
-                           </gmd:dateType>
```
Why is gmd:dateType ending element removed?

jodygarnett replied:
As above, I do not trust that the conflict resolution was working as intended.

jodygarnett replied:
As above this is a whitespace change: ``</gmd:dateType>`` looks to have gained some whitespace during conflict resolution.

ianwallen commented on src/main/plugin/iso19139.ca.HNAP/templates/hnaplevel1-english.xml:
```diff
@@ -372,9 +370,8 @@
-                              <gmd:CI_DateTypeCode codeList="http://nap.geogratis.gc.ca/metadata/register/napMetadataRegister.xml#IC_87"
+                              <gmd:CI_DateTypeCode codeList="https://schemas.metadata.geo.ca/register/napMetadataRegister.xml#IC_87"
                                                    codeListValue="RI_366">creation; création</gmd:CI_DateTypeCode>
-                           </gmd:dateType>
```
Why is gmd:dateType ending element removed?

jodygarnett replied:
conflict resolution whitespace change again.

josegar74 committed:
Fix template / hnaplevel1-english.xml in 3.12.x_transition_schemas_metadata_geo_ca

josegar74 committed:
Remove default namespace from templates

josegar74 committed:
Update codelist urls

josegar74 reviewed (approved):
Applied some small updates and tested locally, the validation looks fine and also the xml produced contains the new urls.

The remove validation can't be tested for now as the server seems not setup.

josegar74 commented on src/sphinx/admin/upgrade.rst:
```diff
@@ -0,0 +1,61 @@
+
+1. Navigate to :menuselection:`Admin console > Tools > Batch process`.
+2. Select records to process, or choose ``All``.
+3. Use :guilabel:`Configure a process` to select the process ``URL replacer``
```
@jodygarnett, this process is not going to replace all the urls, as it is only updating `gmd:URL` elements and `xlink:href` attributes:

https://github.com/geonetwork/core-geonetwork/blob/a1e7663a50e46f916c603707013309a3dd0eca4f/schemas/iso19139/src/main/plugin/iso19139/process/url-host-relocator.xsl#L57-L72

jodygarnett replied:
Do we need to extend url-host-relocator.xsl? Or should we remove this section and rely on the SQL procedure?

josegar74 committed:
FIx SQL command in upgrade instructions

jodygarnett commented:
Thanks for the changes @josegar74 they make sense to me.

I am considering this PR approved ... and just blocked until @bo-lu is in position to update schemas.metadata.geo.ca content.

jodygarnett committed:
Removed incomplete instructions for running bulk update

jodygarnett committed:
Remove unused image on use of bulk upgrade

jodygarnett committed:
Change codeList location to schemas.metadata.geo.ca/register/napMetadataRegister.xml

# Conflicts:
#	src/main/plugin/iso19139.ca.HNAP/templates/hnaplevel1-english.xml
#	src/main/plugin/iso19139.ca.HNAP/templates/hnaplevel1-french.xml

# Conflicts:
#	src/main/plugin/iso19139.ca.HNAP/convert/thesaurus-transformation.xsl

jodygarnett commented:
This is rebased, however I am seeing some issues when validating a produced document:

Testing against local metadata101/schemas hnap:
```
java Validate.java ../hnap/2.3.1/hnap.xsd ~/Downloads/test-record.xml
```
```
/Users/jgarnett/Downloads/test-record.xml validation check: ../hnap/2.3.1/hnap.xsd
/Users/jgarnett/Downloads/test-record.xml: valid
````

When checking agains schemas.metadata.geo.ca (as referenced in downloaded file):
```
java Validate.java --trace ~/Downloads/test-record.xml    
```
```
/Users/jgarnett/Downloads/test-record.xml: validation using schemaLocation
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/metadataApplication.xsd
https://schemas.metadata.geo.ca/2009/gco/gco.xsd
https://schemas.metadata.geo.ca/2009/gco/basicTypes.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/dynamicFeature.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/feature.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryAggregates.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryPrimitives.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryBasic2d.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryBasic0d1d.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/measures.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/units.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/dictionary.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gmlBase.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/basicTypes.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/2009/xlink/xlink.xsd
https://schemas.metadata.geo.ca/2009/xml/xml.xsd
https://schemas.metadata.geo.ca/2009/xml/XMLSchema.dtd --> -//W3C//DTD XMLSCHEMA 200102//EN
https://schemas.metadata.geo.ca/2009/xml/datatypes.dtd --> datatypes
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/temporal.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gmlBase.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/direction.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryBasic0d1d.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/topology.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryComplexes.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryAggregates.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/coverage.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/feature.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/valueObjects.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryBasic0d1d.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/temporal.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/grids.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryBasic0d1d.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/geometryAggregates.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/coordinateReferenceSystems.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/coordinateSystems.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/referenceSystems.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/dictionary.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/datums.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/referenceSystems.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/measures.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/coordinateOperations.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/referenceSystems.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/measures.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/observation.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/feature.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/direction.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/valueObjects.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/temporalReferenceSystems.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/temporalTopology.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/temporal.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/dictionary.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/deprecatedTypes.xsd
https://schemas.metadata.geo.ca/ogc/gml/3.2.1/gml.xsd
https://schemas.metadata.geo.ca/2009/gco/gco.xsd
https://schemas.metadata.geo.ca/2009/gco/gcoBase.xsd
https://schemas.metadata.geo.ca/2009/gco/gco.xsd
https://schemas.metadata.geo.ca/2009/gmd/metadataEntity.xsd
https://schemas.metadata.geo.ca/2009/gmd/spatialRepresentation.xsd
https://schemas.metadata.geo.ca/2009/gss/gss.xsd
https://schemas.metadata.geo.ca/2009/gss/geometry.xsd
https://schemas.metadata.geo.ca/2009/gss/gss.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/referenceSystem.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/extent.xsd
https://schemas.metadata.geo.ca/2009/gts/gts.xsd
https://schemas.metadata.geo.ca/2009/gts/temporalObjects.xsd
https://schemas.metadata.geo.ca/2009/gts/gts.xsd
https://schemas.metadata.geo.ca/2009/gsr/gsr.xsd
https://schemas.metadata.geo.ca/2009/gsr/spatialReferencing.xsd
https://schemas.metadata.geo.ca/2009/gsr/gsr.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/referenceSystem.xsd
https://schemas.metadata.geo.ca/2009/gmd/metadataExtension.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/content.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/metadataApplication.xsd
https://schemas.metadata.geo.ca/2009/gmd/applicationSchema.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/portrayalCatalogue.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/dataQuality.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/identification.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/constraints.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/distribution.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/maintenance.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/citation.xsd
https://schemas.metadata.geo.ca/2009/gmd/freeText.xsd
https://schemas.metadata.geo.ca/2009/gmd/gmd.xsd
https://schemas.metadata.geo.ca/2009/gmd/identification.xsd
https://schemas.metadata.geo.ca/2009/napm/napm.xsd
https://schemas.metadata.geo.ca/2009/napm/napmCharacterString.xsd
test-record.xml (473, 142) ERROR: cvc-complex-type.2.4.a: Invalid content was found starting with element '{"http://www.isotc211.org/2005/gmx":Anchor}'. One of '{"http://www.isotc211.org/2005/gco":CharacterString}' is expected.
/Users/jgarnett/Downloads/test-record.xml: invalid
```

jodygarnett commented:
Checking the record downloaded it has:
```
              <gmd:identifier>
                <gmd:MD_Identifier>
                  <gmd:code>
                    <gmx:Anchor xlink:href="http://localhost:8080/geonetwork/srv/api/registries/vocabularies/external.theme.GC_Core_Subject">geonetwork.thesaurus.external.theme.GC_Core_Subject</gmx:Anchor>
                  </gmd:code>
                </gmd:MD_Identifier>
              </gmd:identifier>
```

The ``gmx/extendedTypes.xsd`` has ``gmx:Anchor`` in substation group for ``gco: CharacterString`` so this should be okay:
```
<xs:element name="Anchor" type="gmx:Anchor_Type" substitutionGroup="gco:CharacterString"/>
```

Looking through the above trace ``gmx`` was not downloaded, adding to the schema location allows the document to validate:
```
xsi:schemaLocation="http://www.isotc211.org/2005/gmx https://schemas.metadata.geo.ca/2009/gmx/gmx.xsd ..."
```

jodygarnett committed:
Include gmx in schemaLocation

josegar74 reviewed (approved):
Changes look fine, thanks.
